### PR TITLE
Support `get_element` from LIST column

### PR DESCRIPTION
**Pull request by isVoid**
Part1 of #8032 

This PR adds retrieval of row data from a `LIST` type column, through adding support to `list_view` specialization of `get_element`. The row data is stored in a scalar object.

Use example:
```
// non-nested LIST column
col = [{1, 2, 3}, {4}]
s = get_element(col, 1); // s is a type erased list_scalar, s._data == int_column{4}

// nested LIST column
col = [[{1, 2}, {3}], [{4}, {}]]
s = get_element(col, 1); // s is a type erased list_scalar, s._data == list_column{{4}, {}}
```

Implementation note:
Depends on `lists::detail::copy_slice` under the hood. Also adds a new `list_scalar` constructor that supports moving external row data to construct a new scalar.

Other included in this PR:
- `is_element_valid_sync(column, i)`, helper function that returns true if `i`th row of `column` is valid.
- `list_scalar` factory functions
- Developer guide for `list_scalar`

### CODE REVIEW

File: cpp/src/copying/get_element.cu
Comment (harrism) [inline cpp/src/copying/get_element.cu:132]: I feel like this code should be in a utility function... I think reading back the 32-bit bitmask word that includes this element to the host and then masking off the right bit would be faster than allocating a device_scalar, launching a kernel, and then copying the result back and syncing.

**Review by harrism — CHANGES_REQUESTED** [review]
Looks fine, just suggest refactoring the single-element is_valid to a utility.

**Review by robertmaynard — APPROVED** [review]

**Comment by ttnghia** [comment]
The behavior seems not right to me. You have a lists column. You access a row in that list column. That row is always a list. Thus, the return list scalar should always have _data that is a list column of one row.

**Comment by isVoid** [comment]
@ttnghia currently `list_scalar` is defined with the top-level list wrapper removed.

For example, in a non-nested list column:
```
// Lists<int>
row0: [1, 2, 3]
row1: [4]
row2: NULL

get_element(0) should return a `list_scalar` with underlying data of INT column {1, 2, 3}
```

For a nested list column:
```
// List<List<int>>
row0:
    row[0][0]: [1, 2, 3]
    row[0][1]: [4, 5]
row1:
    row[1][0]: []
    row[1][1]: NULL

get_element(0) should return `list_scalar` with underlying data of LIST column, which has offset {0, 3, 5} and underlying data of INT column {1, 2, 3, 4, 5}
```

See discussion here: https://github.com/rapidsai/cudf/issues/5887#issuecomment-725050126

**Comment by isVoid** [comment]
I should probably add the definition of `list_scalar` to the developer guide:
https://github.com/rapidsai/cudf/blob/branch-0.20/cpp/docs/DEVELOPER_GUIDE.md#cudfscalar

File: cpp/src/scalar/scalar_factories.cpp
Comment (isVoid) [inline cpp/src/scalar/scalar_factories.cpp:128]: This is not right. The `stream` and `mr` is not being passed to construct the validity data of `list_scalar`.

**Review by robertmaynard — APPROVED** [review]

File: cpp/src/copying/get_element.cu
Comment (vyasr) [reply cpp/src/copying/get_element.cu:132]: There's `bit_is_set` in `utilities/bin.hpp`.

File: cpp/src/bitmask/get_valid.cpp
Comment (harrism) [inline cpp/src/bitmask/get_valid.cpp:40]: You need a stream.synchronize() before the return line, otherwise you could be doing the CPU masking before the copy completes.

File: cpp/include/cudf/detail/get_valid.hpp
Comment (harrism) [inline cpp/include/cudf/detail/get_valid.hpp:30]:
```suggestion
 *
 * @note Synchronizes `stream`.
 *
```

File: cpp/include/cudf/detail/get_valid.hpp
Comment (harrism) [inline cpp/include/cudf/detail/get_valid.hpp:39]: I think it should be called `is_element_valid` (since it returns bool). Perhaps even name it to make the sync obvious.

```suggestion
bool is_element_valid_sync(column_view const& col_view,
```

File: cpp/src/copying/get_element.cu
Comment (harrism) [inline cpp/src/copying/get_element.cu]:
```suggestion
   if (is_element_valid_sync(input, index, stream)) {
```

**Review by harrism — CHANGES_REQUESTED** [review]

**Review by robertmaynard — APPROVED** [review]

**Comment by robertmaynard** [comment]
Approving on behalf of just rapidsai/cudf-cmake-codeowners

File: cpp/tests/copying/get_value_tests.cpp
Comment (nvdbaranec) [inline cpp/tests/copying/get_value_tests.cpp:448]: Some additional tests needed:

- More cases with validity, particularly validity on the inner lists (null int, strings, etc).
- Structs, including list<struct<...>>,   list<struct<list<> ...>>  (this will require the gather change suggested above.

File: cpp/include/cudf/scalar/scalar.hpp
Comment (nvdbaranec) [inline cpp/include/cudf/scalar/scalar.hpp:641]: I think it would be useful to have a constructor that behaves just like the factory function.   Something like:

```
  list_scalar(cudf::column_view const& col,
              size_type row_index,
              rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
              rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
```

**Review by nvdbaranec — CHANGES_REQUESTED** [review]
I think this PR should also include an implementation of `make_column_from_scalar()` for `list_view`

https://github.com/rapidsai/cudf/blob/e6f3f37ae6766cfeef8d8d08474708612cae770d/cpp/src/column/column_factories.cpp#L218

File: cpp/include/cudf/scalar/scalar.hpp
Comment (isVoid) [reply cpp/include/cudf/scalar/scalar.hpp:641]: Does this mean constructing a `list_scalar` from the `row_index`th row of `col`?

**Review by ajschmidt8 — APPROVED** [review]
Approving `ops-codeowner` file changes

File: cpp/include/cudf/scalar/scalar.hpp
Comment (nvdbaranec) [reply cpp/include/cudf/scalar/scalar.hpp:641]: Yes

File: cpp/docs/DEVELOPER_GUIDE.md
Comment (harrism) [inline cpp/docs/DEVELOPER_GUIDE.md:261]:
```suggestion
A `list_scalar` holds the underlying data of a single list. This means the underlying data can be any type
that cudf supports. For example, a `list_scalar` representing a list of integers stores a `cudf::column`
of type `INT32`. A `list_scalar` representing a list of lists of integers stores a `cudf::column` of
type `LIST`, which in turn stores a column of type `INT32`.
```

File: cpp/docs/DEVELOPER_GUIDE.md
Comment (harrism) [inline cpp/docs/DEVELOPER_GUIDE.md:278]: Finish these TODOs before merging?

File: cpp/include/cudf/detail/is_element_valid.hpp
Comment (harrism) [inline cpp/include/cudf/detail/is_element_valid.hpp:29]:
```suggestion
 * Retrieves the validity (NULL or non-NULL) of the specified row from device memory.
```

File: cpp/include/cudf/scalar/scalar.hpp
Comment (harrism) [reply cpp/include/cudf/scalar/scalar.hpp:641]: Then why have the factory?

**Review by harrism — CHANGES_REQUESTED** [review]
Couple doc changes.

File: cpp/docs/DEVELOPER_GUIDE.md
Comment (isVoid) [reply cpp/docs/DEVELOPER_GUIDE.md:278]: This PR did not introduce any new factory functions for `list_scalar`, aka `make_list_scalar(...)`. I'm not quite sure at this point what should the arguments be for this.

File: cpp/tests/copying/get_value_tests.cpp
Comment (isVoid) [reply cpp/tests/copying/get_value_tests.cpp:467]: `list<structs<list<>>>` tests added. `copy_slice` turns out to work just fine with `struct` columns. Also sprinkled some null values into existing fixed-width/string tests.

File: cpp/docs/DEVELOPER_GUIDE.md
Comment (harrism) [reply cpp/docs/DEVELOPER_GUIDE.md:278]: So do we need these TODOs?

**Review by harrism — APPROVED** [review]

File: cpp/docs/DEVELOPER_GUIDE.md
Comment (isVoid) [reply cpp/docs/DEVELOPER_GUIDE.md:278]: The paragraph above suggests a factory function is available for each type of scalars. Maybe instead of adding `TODO` we should rephrase the words above to exclude `list_scalar`.

File: cpp/include/cudf/scalar/scalar.hpp
Comment (nvdbaranec) [reply cpp/include/cudf/scalar/scalar.hpp:638]: We've got these other constructors.  If we have them, why not have another one that provides utility?  Or to put it another way, if we shouldn't have this one, why should we have the other ones?

Comment (nvdbaranec) [reply cpp/include/cudf/scalar/scalar.hpp:638]: We've got these other constructors. If we have them, why not have another one that provides utility? Or to put it another way, if we shouldn't have this one, why should we have the other ones?

**Review by nvdbaranec — CHANGES_REQUESTED** [review]
My comment here still stands.  https://github.com/rapidsai/cudf/pull/8071#discussion_r623942622   We should be using gather.

**Comment by isVoid** [comment]
Added scalar factory methods, and updated developer guide accordingly.

cc @nvdbaranec @harrism

**Comment by isVoid** [comment]
> My comment here still stands. [#8071 (comment)](https://github.com/rapidsai/cudf/pull/8071#discussion_r623942622) We should be using gather.

Discussed offline and confirmed current implementation supports structs.

File: cpp/include/cudf/scalar/scalar.hpp
Comment (isVoid) [reply cpp/include/cudf/scalar/scalar.hpp:638]: Second thoughts: for other scalar types there is always a constructor that constructs the scalar from its "value". `string_scalar(std::string)`, `numeric_scalar(T value)` etc. So it makes sense to have a constructor for `list_scalar` accepting its underlying data.

For the proposed ctor - note this doesn't have to be specific to `list_scalar`, it can actually be implemented in `scalar` base class by punting everything to `get_element`:
```cpp
scalar::scalar(column_view const& col,
          size_type row_index,
          rmm::cuda_stream_view stream,
          rmm::mr::device_memory_resource* mr)
          : scalar(std::move(*detail::get_element(col, row_index, stream, mr).release()))
{}
```
But then this is just another level of indirection.

**Review by nvdbaranec — APPROVED** [review]

**Review by ttnghia — APPROVED** [review]

**Comment by ttnghia** [comment]
It seems that there are some failed tests.

File: cpp/include/cudf/scalar/scalar_factories.hpp
Comment (nvdbaranec) [inline cpp/include/cudf/scalar/scalar_factories.hpp:160]: Sorry to jump in late here.

```suggestion
 * @brief Construct scalar using the given column of elements
```

**Review by nvdbaranec — CHANGES_REQUESTED** [review]

File: cpp/include/cudf/scalar/scalar_factories.hpp
Comment (isVoid) [reply cpp/include/cudf/scalar/scalar_factories.hpp:160]: Better late than never ;)

**Review by nvdbaranec — APPROVED** [review]

**Comment by kkraus14** [comment]
@gpucibot merge

**Pull request merged**
